cmd/experimental/migrate/internal/migrate.go

```diff
@@ -0,0 +1,189 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/avast/retry-go/v4"
+	"github.com/transparency-dev/merkle/rfc6962"
+	"github.com/transparency-dev/trillian-tessera/api"
+	"github.com/transparency-dev/trillian-tessera/api/layout"
+	"github.com/transparency-dev/trillian-tessera/client"
+	"golang.org/x/sync/errgroup"
+	"k8s.io/klog/v2"
+)
+
+// copier controls the migration work.
+type copier struct {
+	// storage is the target we're migrating to.
+	storage    MigrationStorage
+	getEntries client.EntryBundleFetcherFunc
+
+	// sourceSize is the size of the source log.
+	sourceSize uint64
+	// sourceRoot is the root hash of the source log at sourceSize.
+	sourceRoot []byte
+
+	// todo contains work items to be completed.
+	todo chan bundle
+
+	// bundlesMigrated is the number of entry bundles migrated so far.
+	bundlesMigrated atomic.Uint64
+}
+
+// bundle represents the address of an individual entry bundle.
+type bundle struct {
+	Index   uint64
+	Partial uint8
+}
+
+// MigrationStorage describes the required functionality from the target storage driver.
+type MigrationStorage interface {
+	// SetEntryBundle is called to store the provided entry bundle bytes at the given coordinates.
+	SetEntryBundle(ctx context.Context, index uint64, partial uint8, bundle []byte) error
+	// AwaitIntegration should block until the storage driver has received and integrated all outstanding entry bundles implied by sourceSize,
+	// and return the locally calculated root hash.
+	// An error should be returned if there is a problem integrating.
+	AwaitIntegration(ctx context.Context, sourceSize uint64) ([]byte, error)
+	// Size returns the current integrated size of the local tree.
+	Size(ctx context.Context) (uint64, error)
+}
+
+// Migrate starts the work of copying sourceSize entries from the source to the target log.
+//
+// Only the entry bundles are copied as the target storage is expected to integrate them and recalculate the root.
+// This is done to ensure the correctness of both the source log as well as the copy process itself.
+//
+// A call to this function will block until either the copying is done, or an error has occurred.
+// It is an error if the resource copying completes ok but the resulting root hash does not match the provided sourceRoot.
+func Migrate(ctx context.Context, numWorkers int, sourceSize uint64, sourceRoot []byte, getEntries client.EntryBundleFetcherFunc, storage MigrationStorage) error {
+	klog.Infof("Starting migration; source size %d root %x", sourceSize, sourceRoot)
+
+	// TODO store state & resume
```

Comment:
Naked TODO considered harmful. Before submit, can we have this fixed, a name attached, or an issue created and linked?

Reply:
I've removed the TODO - currently if the job crashes and is restarted, it'll pick up from the size of the locally integrated tree. That's likely good enough for now.